### PR TITLE
Workshop survey link

### PR DESCRIPTION
**Pull request by mehalshah**
image

### CODE REVIEW

File: apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx [hunk 1]
```diff
@@ -1,16 +1,64 @@
...
 
   render() {
     return (
       <div>
-        Section for CS Principles and CS Discoveries
+        <br/>
+        <TwoPartBanner
+          leftHalf={(
+            <div
+              style={Object.assign({}, imageStyle, leftStyle, {
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx]: It seems odd that `twoPartBanner.jsx` defines these styles, then they are imported here and applied to children to be passed back into the two part banner. This could be cleaner and more composable. 

Ideally the two part banner wouldn't export styles at all, but rather directly apply them to its children.


Props can be added to existing elements (`leftHalf`, `rightHalf`) with [React.cloneElement](https://facebook.github.io/react/docs/react-api.html#cloneelement). 

Or maybe the TwoPartBanner could construct containing divs for each half with these styles applied:

Something like:
```jsx
<div style={...styles.leftHalf, ...styles.imageStyle}>
  {this.props.leftHalf}
</div>
<div style={...styles.rightHalf, ...styles.textStyle}>
  {this.props.rightHalf}
</div>
```

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  borderBottomLeftRadius: '10px',
+};
+
+const rightStyle = {
+  width: '50%',
+  maxWidth: '50%',
+  float: 'left',
+  borderTopRightRadius: '10px',
+  borderBottomRightRadius: '10px',
+};
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx]: Per [styleguide](https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md#react), these styles should be in a single 'styles' object.

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    return (
+      <div style={containerStyle}>
+        {this.props.leftHalf}
+        {this.props.rightHalf}
+      </div>
+    );
+  }
+});
+
+export {TwoPartBanner, imageStyle, textStyle, leftStyle, rightStyle};
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx]: See above. These styles shouldn't need to be exported. They can either be directly applied to the elements in props (leftHalf, rightHalf), or to container divs. In any case they should stay internal to this component.

File: apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx [hunk 4]
```diff
@@ -1,16 +1,64 @@
...
+    if (this.props.lastWorkshopSurveyUrl) {
+      return (
+        <div id="cspStartSurvey">
+          <h3>
+            Share your feedback
+          </h3>
+          <p>
+            Thank you for completing a Code.org professional learning owrkshop! Please complete this survey so that we
+            may improve your experience.
+          </p>
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx:23]: What about after the survey is completed? This is exposed through `Pd::Enrollment.completed_survey_id` after ~ a minute, or immediately in `PEGASUS_DB[:forms].where(kind: 'PdWorkshopSurvey', source_id: enrollment.id)`

File: dashboard/app/controllers/pd/professional_learning_landing_controller.rb [hunk 5]
```diff
@@ -4,11 +4,18 @@ class Pd::ProfessionalLearningLandingController < ApplicationController
   def index
     # Get the courses that this user teaches
     workshops = Pd::Workshop.enrolled_in_by(current_user)
+    ended_workshops = workshops.in_state(Pd::Workshop::STATE_ENDED)
+
     courses_teaching = workshops.pluck(:course).uniq
```
Comment (aoby) [inline dashboard/app/controllers/pd/professional_learning_landing_controller.rb:9]: outside this PR, but this variable name is misleading. These are courses for workshops this teacher is enrolled in, but not necessarily teaching. It can include workshops that haven't occurred yet, and completed workshops that the teacher has not (yet) taught.

File: dashboard/app/controllers/pd/professional_learning_landing_controller.rb [hunk 6]
```diff
@@ -4,11 +4,18 @@ class Pd::ProfessionalLearningLandingController < ApplicationController
   def index
     # Get the courses that this user teaches
     workshops = Pd::Workshop.enrolled_in_by(current_user)
```
Comment (aoby) [inline dashboard/app/controllers/pd/professional_learning_landing_controller.rb:6]: Note this is counting enrollments, not attendance. This might be ok, but I just wanted to point it out. This will include teachers who registered, but did not show up, for a workshop.

File: apps/test/unit/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSectionTest.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        {...sectionProps}
+      />
+    );
+  };
+
+  it("should show a survey for a user that needs to take a survey", () => {
+    const section = generateSection({coursesTaught: ['CS Principles'], lastWorkshopSurveyUrl: 'url'});
+
+    expect(section.find('#cspStartSurvey')).to.have.length(1);
+    expect(section.find('#cspThanks')).to.have.length(0);
```
Comment (aoby) [inline apps/test/unit/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSectionTest.js]: I'm a little confused here. This doesn't look like a valid state. According to below, `coursesTaught` represents all courses for which this teacher is enrolled, whether or not those workshops have occurred. Surveys are not valid until the workshop has ended. Attempting to navigate to the survey before that will result in a 404.

File: dashboard/app/controllers/pd/professional_learning_landing_controller.rb [hunk 8]
```diff
@@ -4,11 +4,18 @@ class Pd::ProfessionalLearningLandingController < ApplicationController
...
+    courses_completed = ended_workshops.pluck(:course).uniq
+
+    last_enrollment =  Pd::Enrollment.joins(:workshop).where(email: current_user.email).where('pd_enrollments.completed_survey_id IS NULL').order('pd_workshops.ended_at DESC').first
 
     # Link to the certificate
     @landing_page_data = {
-      courses_teaching: courses_teaching
+      courses_teaching: courses_teaching,
+      courses_completed: courses_completed,
+      last_workshop_survey_url: "#{CDO.code_org_url}/pd-workshop-survey/#{last_enrollment.code}"
```
Comment (aoby) [inline dashboard/app/controllers/pd/professional_learning_landing_controller.rb]: In order for this survey_url to be valid, the workshop needs to be ended. The `last_enrollment` query above needs more constraints, and a more descriptive name.

What if there are no enrollments? Line 18 will fail accessing `.code` on nil.

Also this line should be 
```ruby
  last_workshop_survey_url: CDO.code_org_url("/pd-workshop-survey/#{last_enrollment.code})"
```

File: dashboard/db/migrate/20170209010013_add_index_on_pd_enrollment_email.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class AddIndexOnPdEnrollmentEmail < ActiveRecord::Migration[5.0]
+  def change
+    add_index :pd_enrollments, :email
+  end
+end
```
Comment (aoby) [inline dashboard/db/migrate/20170209010013_add_index_on_pd_enrollment_email.rb]: This migration should be extracted into its own PR.

File: dashboard/test/controllers/pd/professional_learning_landing_controller_test.rb [hunk 10]
```diff
@@ -13,7 +13,11 @@ class Pd::ProfessionalLearningLandingControllerTest < ::ActionController::TestCa
       create :pd_enrollment, email: other_teacher.email, workshop: workshop
     end
 
-    create :pd_enrollment, email: @teacher.email, workshop: @csf_workshop
+    @ended_enrollment = create :pd_enrollment, email: @teacher.email, workshop: @csf_workshop
+    create :pd_enrollment, email: @teacher.email, workshop: @csd_workshop
+
+    @csf_workshop.start!
+    @csf_workshop.end!
```
Comment (aoby) [inline dashboard/test/controllers/pd/professional_learning_landing_controller_test.rb]: Note this affects other enrollments in the csf workshop. It _might_ be better to leave those 3 workshops alone and create an additional workshop (or workshops) that are already ended. Your call.

Also note, you can create a workshop that is already ended with [create :pd_ended_workshop](https://github.com/code-dot-org/code-dot-org/blob/c67c4061e5ba0d13f087fe45e56ffed889e877f1/dashboard/test/factories.rb#L537-L542)

File: apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx (hunk 4 above)
Comment (mehalshah) [reply apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx:23]: hmm - now that I think about it, I need it immediately. People will come back to this page after they've completed the survey and I don't want there to be confusion.

The problem is that I don't want to do a big query into the pegasus db every time someone hits this form looking for fresh survey completions. 

Is there a controller for these form responses? What if we updated the Dashboard DB on submission?

File: apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx (hunk 1 above)
Comment (mehalshah) [reply apps/src/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSection.jsx]: I don't really want the two part banner to be constructing the content of the divs. I'd rather the caller say "put A and B in a banner"

Will chat offline

File: dashboard/app/controllers/pd/professional_learning_landing_controller.rb (hunk 6 above)
Comment (mehalshah) [reply dashboard/app/controllers/pd/professional_learning_landing_controller.rb:6]: Yeah, I'm fine with this. If you've enrolled in a CSF class, you're a CSF teacher.

File: dashboard/app/controllers/pd/professional_learning_landing_controller.rb (hunk 5 above)
Comment (mehalshah) [reply dashboard/app/controllers/pd/professional_learning_landing_controller.rb:9]: Ah - I see. Will revisit this

File: apps/test/unit/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSectionTest.js (hunk 7 above)
Comment (mehalshah) [reply apps/test/unit/code-studio/pd/professional_learning_landing/csPrinciplesAndDiscoveriesSectionTest.js]: No - this is valid. The workshop has ended so there's a survey that needs to be taken, hence we'd expect #cspStartSurvey. If the workshop was ended but the survey was already completed, we'd expect #cspThanks

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx [hunk 11]
```diff
@@ -2,59 +2,76 @@
...
-  borderTopLeftRadius: '10px',
-  borderBottomLeftRadius: '10px',
-};
+  leftStyle: {
+    width: '50%',
+    maxWidth: '50%',
+    float: 'left',
+    borderTopLeftRadius: '10px',
+    borderBottomLeftRadius: '10px',
+  },
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx:35]: nit: since these are now contained in a `styles` object, the keys don't need `Style` suffixes. 

I prefer `styles.left` over `styles.leftStyle`

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx [hunk 12]
```diff
@@ -2,59 +2,76 @@
...
+    imagePosition: React.PropTypes.oneOf(['imageLeft', 'imageRight']).isRequired
+  },
+
+  getAlignmentStyle() {
+    return this.props.imagePosition === 'imageLeft' ? styles.leftStyle : styles.rightStyle;
+  },
+
+  renderInterior() {
+    const appliedImageStyle = Object.assign({}, styles.imageStyle, {backgroundImage: this.props.imageUrl}, this.getAlignmentStyle());
+    const appliedTextStyle = Object.assign({}, styles.textStyle, this.getAlignmentStyle());
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx]: Note In ES6, instead of 
```javascript
const appliedImageStyle = Object.assign({}, styles.imageStyle, {backgroundImage: this.props.imageUrl}, this.getAlignmentStyle())
```
you could use the object rest spread syntax:

```ES6
const appliedImageStyle = {
  ...styles.imageStyle, 
  backgroundImage: this.props.imageUrl,
  ...this.getAlignmentStyle()
}

```

File: dashboard/test/controllers/pd/professional_learning_landing_controller_test.rb [hunk 13]
```diff
@@ -13,7 +13,10 @@ class Pd::ProfessionalLearningLandingControllerTest < ::ActionController::TestCa
       create :pd_enrollment, email: other_teacher.email, workshop: workshop
     end
 
-    create :pd_enrollment, email: @teacher.email, workshop: @csf_workshop
+    @ended_enrollment = create :pd_enrollment, email: @teacher.email, workshop: @csf_workshop, survey_sent_at: Date.today + 5.days
+    other_enrollment = create :pd_enrollment, email: @teacher.email, workshop: @csd_workshop, survey_sent_at: Date.today + 2.days
```
Comment (aoby) [inline dashboard/test/controllers/pd/professional_learning_landing_controller_test.rb:17]: This other_enrollment is not in a normal state with the workshop not ended but a survey sent. It may not affect the test, but it would be cleaner and more representative if the `csd_workshop` was also a `pd_ended_workshop`

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx (hunk 11 above)
Comment (mehalshah) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx]: Agreed

File: apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx [hunk 14]
```diff
@@ -26,15 +26,15 @@ const styles = {
     padding: '20px'
   },
 
-  leftStyle: {
+  left: {
```
Comment (aoby) [inline apps/src/code-studio/pd/professional_learning_landing/twoPartBanner.jsx]: the other styles also don't need the suffix: `styles.containerStyle` etc.
